// ldes-client/dist/lib/strategy/unordered.js
import debug from "debug";
const log = debug("strategy");
export class UnorderedStrategy {
    manager;
    fetcher;
    notifier;
    inFlight = 0;
    fetchNotifier;
    memberNotifier;
    modulator;
    cacheList = [];
    polling;
    pollInterval;
    cancled = false;
    constructor(memberManager, fetcher, notifier, modulatorFactory, polling, pollInterval) {
        this.pollInterval = pollInterval;
        this.notifier = notifier;
        this.manager = memberManager;
        this.fetcher = fetcher;
        this.polling = polling;
        const fetchLogger = log.extend("fetch");
        // Callbacks for the fetcher
        // - seen: the strategy wanted to fetch an uri, but it was already seen
        //         so one fetch request is terminated, inFlight -= 1
        // - pageFetched: a complete page is fetched and the relations have been extracted
        //         start member extraction
        // - relationFound: a relation has been found, inFlight += 1 and put it in the queue
        this.fetchNotifier = {
            error: (error) => {
                fetchLogger("error %o", error);
                this.notifier.error(error, {});
            },
            scheduleFetch: (node) => {
                this.cacheList.push(node);
                this.notifier.mutable({}, {});
            },
            pageFetched: (page, { index }) => {
                fetchLogger("Paged fetched %s", page.url);
                this.handleFetched(page, index);
            },
            relationFound: ({ from, target }) => {
                from.expected.push(target.node);
                this.inFlight += 1;
                this.modulator.push({ target: target.node, expected: [from.target] });
            },
        };
        const memberLogger = log.extend("member");
        // Callbacks for the member extractor
        // - done: all members have been extracted, we are finally done with a page inFlight -= 1
        // - extracted: a member has been found, yeet it
        this.memberNotifier = {
            error: (error) => {
                this.notifier.error(error, {});
            },
            done: () => {
                memberLogger("Members on page done");
                this.inFlight -= 1;
                this.checkEnd();
                this.notifier.fragment({}, {});
            },
            extracted: (mem) => this.notifier.member(mem, {}),
        };
        this.modulator = modulatorFactory.create("fetcher", [], {
            ready: ({ item, index }) => this.fetcher.fetch(item, { index }, this.fetchNotifier),
        });
    }
    start(url) {
        const logger = log.extend("start");
        this.inFlight = this.modulator.length();
        if (this.inFlight < 1) {
            this.inFlight = 1;
            this.modulator.push({ target: url, expected: [] });
            logger("Nothing in flight, adding start url");
        }
        else {
            logger("Things are already inflight, not adding start url");
        }
    }
    cancel() {
        this.cancled = true;
    }
    handleFetched(page, index) {
        this.modulator.finished(index);
        this.manager.extractMembers(page, {}, this.memberNotifier);
    }
    checkEnd() {
        if (this.cancled)
            return;
        if (this.inFlight == 0) {
            if (this.polling) {
                setTimeout(() => {
                    if (this.cancled)
                        return;
                    this.notifier.pollCycle({}, {});
                    const cl = this.cacheList.slice();
                    this.cacheList = [];
                    for (let cache of cl) {
                        this.inFlight += 1;
                        this.modulator.push(cache);
                    }
                }, this.pollInterval || 1000);
            }
            else {
                log("Closing the notifier, polling is not set");
                this.cancled = true;
                this.notifier.close({}, {});
            }
        }
    }
}
